interfaces/udev/backend.go

```diff
@@ -73,66 +80,104 @@ func (b *Backend) Setup(appSet *interfaces.SnapAppSet, opts interfaces.Confineme
 	snapName := appSet.InstanceName()
 	spec, err := repo.SnapSpecification(b.Name(), appSet)
 	if err != nil {
-		return fmt.Errorf("cannot obtain udev specification for snap %q: %s", snapName, err)
+		return fmt.Errorf("cannot obtain udev specification for snap %q: %w", snapName, err)
 	}
-	content := b.deriveContent(spec.(*Specification))
-	subsystemTriggers := spec.(*Specification).TriggeredSubsystems()
+
+	udevSpec := spec.(*Specification)
+	content := b.deriveContent(udevSpec)
+	subsystemTriggers := udevSpec.TriggeredSubsystems()
 
 	dir := dirs.SnapUdevRulesDir
 	if err := os.MkdirAll(dir, 0755); err != nil {
-		return fmt.Errorf("cannot create directory for udev rules %q: %s", dir, err)
+		return fmt.Errorf("cannot create directory for udev rules %q: %w", dir, err)
 	}
 
 	rulesFilePath := snapRulesFilePath(snapName)
+	selfManageDeviceCgroupPath := snapDeviceCgroupSelfManageFilePath(snapName)
 
-	if len(content) == 0 {
+	if len(content) == 0 || udevSpec.ControlsDeviceCgroup() {
 		// Make sure that the rules file gets removed when we don't have any
 		// content and exists.
 		err = os.Remove(rulesFilePath)
-		if err != nil && !os.IsNotExist(err) {
+		if err != nil && !errors.Is(err, fs.ErrNotExist) {
 			return err
 		} else if err == nil {
 			// FIXME: somehow detect the interfaces that were
 			// disconnected and set subsystemTriggers appropriately.
 			// ATM, it is always going to be empty on disconnect.
-			return b.reloadRules(subsystemTriggers)
+			if err := b.reloadRules(subsystemTriggers); err != nil {
+				return err
+			}
 		}
-		return nil
 	}
 
-	var buffer bytes.Buffer
-	buffer.WriteString("# This file is automatically generated.\n")
-	if (opts.DevMode || opts.Classic) && !opts.JailMode {
-		buffer.WriteString("# udev tagging/device cgroups disabled with non-strict mode snaps\n")
+	if err := os.MkdirAll(dirs.SnapCgroupPolicyDir, 0755); err != nil {
+		return fmt.Errorf("cannot create directory for cgroup flags: %w", err)
 	}
```

Comment:
should this go closer to where we touch this files?

Reply:
I've actually moved this closer to where we MkdirAll the udev rules directory, so that all prerequisites are satisfied before we reach the code generating the rules.